keyboards/aplusx_pulsar/xboard_qs/keymaps/via/keymap.c

```diff
@@ -0,0 +1,400 @@
+        KC_LCTL, KC_LALT, KC_LGUI,                   KC_SPC,                       KC_RGUI,  MO(MAC_FN), KC_RALT, KC_RCTL,           KC_LEFT, KC_DOWN, KC_RGHT),
+
+    [MAC_FN]   = LAYOUT(/* Layer 3 ; MAC Function Layer */
+                                                                                                                                KC_PC1, KC_PC2, RGB_MOD,
+        INIT_3S,      KC_F1,   KC_F2,   KC_F3,   KC_F4,     KC_F5,   KC_F6,   KC_F7,   KC_F8,     KC_F9,   KC_F10,  KC_F11,  KC_F12, KC_TMODE, KC_F13, KC_F14, KC_F15,
+        DM_RSTP, DM_REC1, DM_REC2, DM_PLY1, DM_PLY2, _______, _______, _______, _______, _______, _______, NK_OFF,   NK_ON,  RGB_TOG,          RGB_SAI, RGB_HUI, RGB_MOD,
+        _______, KC_BTN1, KC_MS_U, KC_BTN2, KC_WH_U, _______, _______, _______, _______, _______, _______, _______, _______, _______,          RGB_SAD, RGB_HUD, RGB_RMOD,
+        _______, KC_MS_L, KC_MS_D, KC_MS_R, KC_WH_D, _______, _______, _______, _______, _______, _______, _______,          _______,
+        _______,         KC_BTN4, KC_BTN3, KC_BTN5, _______, KC_S_LED, _______, _______, _______, _______, _______,          _______,                   RGB_VAI,
+        _______, _______, _______,                         _______,                               _______, _______, _______,  KC_TPC,          RGB_SPD, RGB_VAD, RGB_SPI)
+        };
+//-----------------------------------------------------------------
+#if defined(ENCODER_MAP_ENABLE)
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
+    [WIN_BASE] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [WIN_FN]   = { ENCODER_CCW_CW(RGB_VAD, RGB_VAI) },
+    [MAC_BASE] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [MAC_FN]   = { ENCODER_CCW_CW(RGB_VAD, RGB_VAI) }
+};
+#endif // ENCODER_MAP_ENABLE
+//-----------------------------------------------------------------
+#define HCS(report) host_consumer_send(record->event.pressed ? report : 0); return false
+#define HSS(report) host_system_send(record->event.pressed ? report : 0); return false
+
+static bool win_key_locked = false;
+static uint8_t mac_keycode[4] = {KC_LOPT, KC_ROPT, KC_LCMD, KC_RCMD};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+
+    switch (keycode) {
+        case INIT_3S:
+            if (record->event.pressed){
+                INIT3S_timer = timer_read();
+                INIT3S_on    = true;
+                return false;
+            }
+            else{
+                INIT3S_on   = false;
+                return false;
+            }
+            break;
+
+        case KC_TGUI:
+            if (record->event.pressed) {                // Toggle GUI lock on key press
+                win_key_locked = !win_key_locked;
+            }
+            break;
+
+        case KC_LGUI:
+            if (win_key_locked) { return false; }
+            else{ return true; }
+
+        case KC_MISSION_CONTROL:
```

Comment:
Some of these already exist in code, and should not be reimplemented at the keymap level